File: src/main/java/jugglinglab/path/Path.java
//
// Path.java
//
// This is the base class for all Path types in Juggling Lab. A Path describes
// the movement of an object during the time between throw and catch.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.path;

import java.util.*;
import jugglinglab.util.*;

public abstract class Path {
  static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

  protected double start_time;
  protected double end_time;
  protected Coordinate start_coord;
  protected Coordinate end_coord;

  // the built-in path types
  public static final String[] builtinPaths = {"Toss", "Bounce"};

  // Create a new path of the given type.

  public static Path newPath(String type) throws JuggleExceptionUser {
    if (type == null) {
      throw new JuggleExceptionUser("Path type not specified");
    }

    if (type.equalsIgnoreCase("toss")) {
      return new TossPath();
    } else if (type.equalsIgnoreCase("bounce")) {
      return new BouncePath();
    }

    throw new JuggleExceptionUser("Path type '" + type + "' not recognized");
  }

  public void setStart(Coordinate position, double time) {
    start_coord = position;
    start_time = time;
  }

  public void setEnd(Coordinate position, double time) {
    end_coord = position;
    end_time = time;
  }

  public double getStartTime() {
    return start_time;
  }

  public double getEndTime() {
    return end_time;
  }

  public double getDuration() {
    return (end_time - start_time);
  }

  // Minimum duration is nonzero for certain throw types, e.g., a double
  // bounce throw.

  public double getMinDuration() {
    return 0.0;
  }

  public void translateTime(double deltat) {
    start_time += deltat;
    end_time += deltat;
  }

  // For screen layout.

  public Coordinate getMax() {
    return getMax2(start_time, end_time);
  }

  public Coordinate getMin() {
    return getMin2(start_time, end_time);
  }

  public Coordinate getMax(double begin, double end) {
    if (end < start_time || begin > end_time) return null;
    return getMax2(begin, end);
  }

  public Coordinate getMin(double begin, double end) {
    if (end < start_time || begin > end_time) {
      return null;
    }
    return getMin2(begin, end);
  }

  // Utility for getMax/getMin.

  protected Coordinate check(Coordinate result, double t, boolean findmax) {
    Coordinate loc = new Coordinate(0, 0, 0);
    getCoordinate(t, loc);
    if (findmax) {
      result = Coordinate.max(result, loc);
    } else {
      result = Coordinate.min(result, loc);
    }
    return result;
  }

  // string indicating the type of path
  public abstract String getType();

  // used for defining the path in the UI (EditLadderDiagram)
  public abstract ParameterDescriptor[] getParameterDescriptors();

  // defines the path from a config string
  public abstract void initPath(String st) throws JuggleExceptionUser;

  public abstract void calcPath() throws JuggleExceptionInternal;

  // for hand layout purposes, only valid after calcPath()
  public abstract Coordinate getStartVelocity();

  public abstract Coordinate getEndVelocity();

  // only valid after calcPath()
  public abstract void getCoordinate(double time, Coordinate newPosition);

  // for hand layout, only valid after calcPath()
  protected abstract Coordinate getMax2(double begin, double end);

  protected abstract Coordinate getMin2(double begin, double end);
}

File: src/main/kotlin/jugglinglab/path/Path.kt
//
// Path.kt
//
// This is the base class for all Path types in Juggling Lab. A Path describes
// the movement of an object from one point in spacetime to another.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.path

import jugglinglab.util.Coordinate
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.ParameterDescriptor

abstract class Path {
    // Coordinate and time of path start
    protected var startCoord: Coordinate? = null
    protected var startTime: Double = 0.0

    // Coordinate and time of path end
    protected var endCoord: Coordinate? = null
    protected var endTime: Double = 0.0

    //--------------------------------------------------------------------------
    // Methods to define the Path
    //--------------------------------------------------------------------------

    fun setStart(position: Coordinate, time: Double) {
        startCoord = position
        this.startTime = time
    }

    fun setEnd(position: Coordinate, time: Double) {
        endCoord = position
        this.endTime = time
    }

    @Throws(JuggleExceptionUser::class)
    abstract fun initPath(st: String?)

    // Must be called after above path parameters are set, before querying for
    // path coordinates
    @Throws(JuggleExceptionInternal::class)
    abstract fun calcPath()

    // Utility method
    @Suppress("unused")
    fun translateTime(deltat: Double) {
        this.startTime += deltat
        this.endTime += deltat
    }

    //--------------------------------------------------------------------------
    // Querying properties of the path
    //--------------------------------------------------------------------------

    // String indicating the type of path
    abstract val type: String

    val duration: Double
        get() = (this.endTime - this.startTime)

    // Minimum duration between `startCoord` and `endCoord`, for a path of the
    // given type
    abstract val minDuration: Double

    // Parameters for defining the path in the UI (e.g., EditLadderDiagram)
    abstract val parameterDescriptors: Array<ParameterDescriptor>

    // Calculated velocity at the start and end of the path, for hand layout
    abstract val startVelocity: Coordinate
    abstract val endVelocity: Coordinate

    // Calculated path coordinate at time `time`
    abstract fun getCoordinate(time: Double, newPosition: Coordinate)

    //--------------------------------------------------------------------------
    // Max and min coordinates over a range of times, used for layout
    //--------------------------------------------------------------------------

    // Path max/min coordinates over time interval [time1, time2]
    protected abstract fun getMax2(time1: Double, time2: Double): Coordinate?
    protected abstract fun getMin2(time1: Double, time2: Double): Coordinate?

    // Max/min over the entire path duration
    val max: Coordinate?
        get() = getMax2(this.startTime, this.endTime)
    val min: Coordinate?
        get() = getMin2(this.startTime, this.endTime)

    // Path max/min over [time1, time2], but clipped to `null` when the time is
    // out of range
    fun getMax(time1: Double, time2: Double): Coordinate? {
        if (time2 < this.startTime || time1 > this.endTime) return null
        return getMax2(time1, time2)
    }
    fun getMin(time1: Double, time2: Double): Coordinate? {
        if (time2 < this.startTime || time1 > this.endTime) return null
        return getMin2(time1, time2)
    }

    // Utility for getMax2/getMin2
    protected fun check(result: Coordinate?, t: Double, findmax: Boolean): Coordinate {
        val loc = Coordinate(0.0, 0.0, 0.0)
        getCoordinate(t, loc)

        val res = if (result == null) {
            loc
        } else if (findmax) {
            Coordinate.max(result, loc)
        } else {
            Coordinate.min(result, loc)
        }
        return res!!
    }

    companion object {
        // List of the built-in path types
        @JvmField
        val builtinPaths: Array<String> = arrayOf<String>("Toss", "Bounce")

        // Factory method to create paths
        @JvmStatic
        @Throws(JuggleExceptionUser::class)
        fun newPath(type: String): Path {
            if (type.equals("toss", ignoreCase = true)) {
                return TossPath()
            } else if (type.equals("bounce", ignoreCase = true)) {
                return BouncePath()
            }
            throw JuggleExceptionUser("Path type '$type' not recognized")
        }
    }
}
